### ISSUE TITLE
IOS add a photo library selection changed callback

### ISSUE THREAD

**Issue by luizppa** [zoontek/react-native-permissions#689]
### Why it is needed?

I have noticed the lack of this feature during [a discussion on react-native-image-crop-picker](https://github.com/ivpusic/react-native-image-crop-picker/issues/1569). When the app displays library photos for the user and allows him to change the limited selection, as far as I could tell, there is no way to detect when the selection change has finished in order to update the displayed photos. This leads to a poor user experience and to potential bugs (like trying to access a photo that your app no longer has access to).

### Possible implementation

I have no experience with IOS native development, so I can not tell if there is any technical barriers to doing so, but I believe we should have a callback fired when the user finishes editing the allowed photo selection. As I mentioned, I would not know how to do it at first, but I can try making a PR if someone would point me on the right direction.

The result could be something like bellow.

### Code sample

```js
import {openLimitedPhotoLibraryPicker} from 'react-native-permissions';

openLimitedPhotoLibraryPicker()
  .catch(() => {
    console.warn('Cannot open photo library picker');
  })
  .then(() => {
    console.log('Limited photo selection changed');
  });
```

**Comment by mikehardy** [comment]
I was just curious so I looked through this, it looks like if you ask for permission again (does this library permit that?) the user can manage the limited set again so a call to action saying "Add Photos?" that re-requests limited permission might do it?

**Comment by luizppa** [comment]
@mikehardy it is not really a matter of when to prompt the user to manage the limited selection but rather when to know he has finished doing so. 

Take this example scenario:
- The user has photos A, B and C on his library, and currently has permitted the app access to photos A and B.
- He navigates to a screen where he is presented with his photo library (only A and B, since that is all we can access) and is supposed to select a photo for upload. This screen also has an option to modify the limited photo selection.
- He chooses to modify this selection, deselects photo B and presses "done".
- Since there is no callback to execute when the selection change has finished, no new render will be triggered if the user remains on the same screen, therefore we will still be displaying photo B.
- If the user then chooses to upload photo B, an error will occur where our app tries to access a file it is not allowed to have.

Likewise, if the user adds photo C to the selection, it will not be displayed until he navigates to another screen and go back again.

**Comment by mikehardy** [comment]
There's no callback in the API that I know of. How do the AppState APIs built-in to react-native react? Might be a foreground to background transition and back.

**Comment by luizppa** [comment]
Hi @mikehardy! Thanks for your fast response and I am sorry for the delay. We are having some problems with this feature on the simulator running on apple silicon, which is my case, so I had to wait for support from my co-worker who has an intel chip to try your suggestion.

Unfortunately it did not work, the app does not seem to transition neither to background nor foreground during the limited selection management.

**Comment by luizppa** [comment]
I did some digging and it looks like this is the code that opens the file picker using the PHPhotoLibrary class: https://github.com/zoontek/react-native-permissions/blob/master/ios/PhotoLibrary/RNPermissionHandlerPhotoLibrary.m#L62

Looking through the class docs, it would seem that we can use this version of the method to fire a callback: https://developer.apple.com/documentation/photokit/phphotolibrary/3752108-presentlimitedlibrarypicker

Like I said, I am not used to iOS development, so I am not completely sure this makes sense. If it does, I could open a PR.

**Comment by mikehardy** [comment]
Great find!

My previous statement "There's no callback in the API that I know of" is invalidated by that second link I missed it when looking at API docs. It looks like that API would make sense and work for the use case.

**Comment by zoontek** [comment]
@mikehardy It doesn't exist for iOS 14.x (where the limited feature has been added first). So the solution is partial here.

**Comment by mikehardy** [comment]
A fair point, but one of the nice things about iOS people is they upgrade really quickly, so it covers most immediately, and almost all "relatively quickly". The alternative being "it's not possible" I think.

Could maybe do an `@available` guard for ios 15.0 ([example, but 14.0](https://github.com/zoontek/react-native-permissions/blob/37d732e8333f1a653300292e82d08decaf9e3995/ios/PhotoLibrary/RNPermissionHandlerPhotoLibrary.m#L21-L25)) to use the version with completion handler, and just document clearly the limitation on iOS < 15 when it can't use that API and has to fallback on the no-completion handler case

In other words, arguing that incremental progress is valuable since I don't think there's any other way to  handle the use case otherwise

**Comment by zoontek** [comment]
@mikehardy I think we could listen for `PHFetchResultChangeDetails` (https://developer.apple.com/documentation/photokit/phfetchresultchangedetails) on iOS < 15, will try to implement it.

**Comment by mechazod** [comment]
@luizppa I think that is not possible, I suggest you make a listener if there's photo library change.

**Comment by anisoft89** [comment]
will you add this feature to detect that window closed?

**Comment by jenipharachel** [comment]
@luizppa How did you work around this issue in your case?

**Comment by jenipharachel** [comment]
> @mikehardy I think we could listen for `PHFetchResultChangeDetails` (https://developer.apple.com/documentation/photokit/phfetchresultchangedetails) on iOS < 15, will try to implement it.

@zoontek I'm aware that you'd be having a lot on your plate. But since I don't have experience working with native iOS, would this feature be available anytime soon? I'm also open to other ideas to handle this issue from my end.

**Comment by zoontek** [comment]
@jenipharachel This is the kind of feature I don't personally need, but I'm available for freelancing work if your company really need it! You can contact me via email 🙂

**Comment by luizppa** [comment]
@jenipharachel I made a fork with a callback. I'm not very familiar with native iOS and React Bridge, so it's not the fancy listeners that were suggested, but it is a good enough workaround if you are not too concerned with older iOS versions.

Edit: the fact that this is kinda of a partial solution is the reason I didn't open a PR

Here is the commit link, it also has an example: https://github.com/luizppa/react-native-permissions/commit/37c8c13c197c2594cd57aedb51a97c28b8da45c7

**Comment by zoontek** [comment]
Just discovered that a similar job has to be done for Android now (https://developer.android.com/about/versions/14/changes/partial-photo-video-access).
I put the feature request on https://polar.sh/zoontek, if companies are interested in it.

**Comment by zoontek** [comment]
@luizppa (and others who want to try), this is available in latest v4 beta: https://github.com/zoontek/react-native-permissions/pull/808/commits/b72d3d39d9d3325d0526d4094ea724fc059a9099 (and it also works on iOS 14, using a `NSTimer` trick).

To try it, `yarn add react-native-permissions@next`. Note that I renamed the function to `openPhotoPicker` (to unify naming with the future Android implem) and that you might need to add `PhotosUI.framework` to the app target, as it appears there a regression on Xcode 15 (see https://developer.apple.com/forums/thread/737847?answerId=770295022#770295022)

**Comment by zoontek** [comment]
Closed as [v4.0.0 is out](https://github.com/zoontek/react-native-permissions/releases/tag/4.0.0).
